hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java

```diff
@@ -934,13 +939,61 @@ private Result increment(final HRegion region, final OperationQuota quota,
         builder.addResultOrException(resultOrExceptionBuilder.build());
       }
     }
+    // do multiget in parallel
+    if (getCtxs != null && !getCtxs.isEmpty()) {
+      doParallelGet(getCtxs, cellsToReturn, builder);
+    }
     // Finish up any outstanding mutations
     if (!CollectionUtils.isEmpty(mutations)) {
       doNonAtomicBatchOp(builder, region, quota, mutations, cellScanner, spaceQuotaEnforcement);
     }
     return cellsToReturn;
   }
 
+  private void doParallelGet(List<GetContext> getCtxs, List<CellScannable> cellsToReturn,
+      RegionActionResult.Builder builder) throws ServiceException {
+    ResultOrException.Builder resultOrExceptionBuilder = null;
+    CountDownLatch latch = new CountDownLatch(getCtxs.size());
+    List<GetActionHandler> handlers = new ArrayList<>(getCtxs.size());
+    for (GetContext getCtx : getCtxs) {
+      GetActionHandler handler = new GetActionHandler(getCtx, latch);
+      this.regionServer.executorService.submit(handler);
+      handlers.add(handler);
+    }
+    try {
+      latch.await();
```

Comment:
The countdown() happens in the handler. What if the countDown happens even before the await() is called - where the isAboveQuota() happens?

Reply:
Sorry Sir, I don't quite understand the impact you worry about. you mean isAboveQuota will not happen in that case?

Reply:
My concern was since the handler is already submitted to the executor, before even the await() happens there could be a chance that the countdown() could be called. Seems calling countDown() without await() is not a problem. But the reverse could be a problem. Will that happen?

Reply:
Read the javadoc of CountDownLatch, seems the Driver2 Sample usage are also in the same way. It doesn't matter which method (countDown & await) is executed first ?

Reply:
If await() happens after countDown() - then the thread that calls await() will hang till it is interrupted. So just ensure that case does not happen here.

Reply:
Got it.